File: app/src/main/java/com/demo/railbridge/sdk/MockRailSdkAdapter.java
package com.demo.railbridge.sdk;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class MockRailSdkAdapter implements RailPlusSdkAdapter {

    private final Backend backend;
    private final ScenarioController scenarioController;
    private final Map<String, Integer> requestAttempts = new ConcurrentHashMap<>();

    public MockRailSdkAdapter(MockRailSdk mockRailSdk) {
        this(new LiveBackend(mockRailSdk), new ScenarioController());
    }

    public MockRailSdkAdapter(MockRailSdk mockRailSdk, ScenarioController scenarioController) {
        this(new LiveBackend(mockRailSdk), scenarioController);
    }

    MockRailSdkAdapter(Backend backend) {
        this(backend, new ScenarioController());
    }

    MockRailSdkAdapter(Backend backend, ScenarioController scenarioController) {
        this.backend = backend;
        this.scenarioController = scenarioController == null ? new ScenarioController() : scenarioController;
    }

    @Override
    public void initialize(Callback<Boolean> callback) {
        backend.initialize(callback);
    }

    @Override
    public void requestCharge(String cardId, int amount, Callback<ChargeResult> callback) {
        handleScenarioAwareRequest(
                buildRequestKey("requestCharge", cardId, amount),
                wrappedCallback -> backend.requestCharge(cardId, amount, wrappedCallback),
                callback
        );
    }

    @Override
    public void getBalance(String cardId, Callback<BalanceSnapshot> callback) {
        handleScenarioAwareRequest(
                buildRequestKey("getBalance", cardId, null),
                wrappedCallback -> backend.getBalance(cardId, wrappedCallback),
                callback
        );
    }

    @Override
    public SdkStatusSnapshot getStatus() {
        return backend.getStatus();
    }

    @Override
    public void shutdown() {
        backend.shutdown();
    }

    public ScenarioController getScenarioController() {
        return scenarioController;
    }

    private <T> void handleScenarioAwareRequest(
            String requestKey,
            BackendRunner<T> backendRunner,
            Callback<T> callback
    ) {
        ScenarioOutcome outcome = nextOutcome(requestKey);

        if (!outcome.emitsCallback()) {
            try {
                backendRunner.run(new NoOpCallback<T>());
            } finally {
                if (outcome.shouldClearAttemptsAfterHandling()) {
                    clearAttempts(requestKey);
                }
            }
            return;
        }

        if (!outcome.usesBackendSuccessPath()) {
            try {
                callback.onError(outcome.getErrorCode());
            } finally {
                if (outcome.shouldClearAttemptsAfterHandling()) {
                    clearAttempts(requestKey);
                }
            }
            return;
        }

        backendRunner.run(new ScenarioCallback<>(callback, requestKey, outcome));
    }

    private ScenarioOutcome nextOutcome(String requestKey) {
        int attempt = requestAttempts.merge(requestKey, 1, Integer::sum);
        return ScenarioOutcome.forAttempt(scenarioController.getActivePreset(), attempt);
    }

    private void clearAttempts(String requestKey) {
        requestAttempts.remove(requestKey);
    }

    private String buildRequestKey(String method, String cardId, Integer amount) {
        StringBuilder builder = new StringBuilder(method)
                .append('|')
                .append(cardId == null ? "" : cardId.trim());
        if (amount != null) {
            builder.append('|').append(amount);
        }
        return builder.toString();
    }

    interface Backend {
        void initialize(Callback<Boolean> callback);

        void requestCharge(String cardId, int amount, Callback<ChargeResult> callback);

        void getBalance(String cardId, Callback<BalanceSnapshot> callback);

        SdkStatusSnapshot getStatus();

        void shutdown();
    }

    private interface BackendRunner<T> {
        void run(Callback<T> callback);
    }

    private final class ScenarioCallback<T> implements Callback<T> {
        private final Callback<T> callback;
        private final String requestKey;
        private final ScenarioOutcome outcome;

        private ScenarioCallback(Callback<T> callback, String requestKey, ScenarioOutcome outcome) {
            this.callback = callback;
            this.requestKey = requestKey;
            this.outcome = outcome;
        }

        @Override
        public void onSuccess(T result) {
            try {
                callback.onSuccess(result);
                if (outcome.isDuplicateCallback()) {
                    callback.onSuccess(result);
                }
            } finally {
                clearAttempts(requestKey);
            }
        }

        @Override
        public void onError(SdkErrorCode errorCode) {
            try {
                callback.onError(errorCode);
            } finally {
                clearAttempts(requestKey);
            }
        }
    }

    private static final class NoOpCallback<T> implements Callback<T> {
        @Override
        public void onSuccess(T result) {
        }

        @Override
        public void onError(SdkErrorCode errorCode) {
        }
    }

    private static final class LiveBackend implements Backend {

        private final MockRailSdk mockRailSdk;

        private LiveBackend(MockRailSdk mockRailSdk) {
            this.mockRailSdk = mockRailSdk;
        }

        @Override
        public void initialize(Callback<Boolean> callback) {
            mockRailSdk.initialize(new MockRailSdk.SdkCallback<Boolean>() {
                @Override
                public void onSuccess(Boolean result) {
                    callback.onSuccess(result);
                }

                @Override
                public void onError(SdkErrorCode errorCode) {
                    callback.onError(errorCode);
                }
            });
        }

        @Override
        public void requestCharge(String cardId, int amount, Callback<ChargeResult> callback) {
            mockRailSdk.charge(cardId, amount, new MockRailSdk.SdkCallback<ChargeResult>() {
                @Override
                public void onSuccess(ChargeResult result) {
                    callback.onSuccess(result);
                }

                @Override
                public void onError(SdkErrorCode errorCode) {
                    callback.onError(errorCode);
                }
            });
        }

        @Override
        public void getBalance(String cardId, Callback<BalanceSnapshot> callback) {
            mockRailSdk.getBalance(cardId, new MockRailSdk.SdkCallback<MockRailSdk.BalanceResult>() {
                @Override
                public void onSuccess(MockRailSdk.BalanceResult result) {
                    callback.onSuccess(new BalanceSnapshot(
                            result.getCardId(),
                            result.getBalance(),
                            result.getTimestamp()
                    ));
                }

                @Override
                public void onError(SdkErrorCode errorCode) {
                    callback.onError(errorCode);
                }
            });
        }

        @Override
        public SdkStatusSnapshot getStatus() {
            return new SdkStatusSnapshot(mockRailSdk.isInitialized(), mockRailSdk.getSdkVersion());
        }

        @Override
        public void shutdown() {
            mockRailSdk.shutdown();
        }
    }
}

File: ios/RailBridgeIOS/RailBridgeIOS/SDK/MockRailSdkAdapter.swift
import Foundation

final class MockRailSdkAdapter: RailPlusSdkAdapter {
    private let queue = DispatchQueue(label: "com.demo.railbridge.ios.mock-sdk", qos: .userInitiated)
    private let scenarioController: ScenarioController
    private var requestAttempts: [String: Int] = [:]
    private var initialized = false

    init(scenarioController: ScenarioController = ScenarioController()) {
        self.scenarioController = scenarioController
    }

    func initialize(completion: @escaping (Bool) -> Void) {
        queue.asyncAfter(deadline: .now() + 0.05) { [weak self] in
            self?.initialized = true
            completion(true)
        }
    }

    func requestCharge(
        cardId: String,
        amount: Int,
        completion: @escaping (Result<ChargeResult, SdkErrorCode>) -> Void
    ) {
        handleScenarioAwareRequest(
            requestKey: buildRequestKey(method: "requestCharge", cardId: cardId, amount: amount),
            backend: { callback in
                self.queue.asyncAfter(deadline: .now() + 0.15) {
                    let result = ChargeResult(
                        transactionId: "IOS-TXN-\(Int(Date().timeIntervalSince1970 * 1000))",
                        amount: amount,
                        balance: max(0, 38947 - amount),
                        timestamp: ISO8601DateFormatter().string(from: Date())
                    )
                    callback(.success(result))
                }
            },
            completion: completion
        )
    }

    func getBalance(
        cardId: String,
        completion: @escaping (Result<BalanceSnapshot, SdkErrorCode>) -> Void
    ) {
        handleScenarioAwareRequest(
            requestKey: buildRequestKey(method: "getBalance", cardId: cardId, amount: nil),
            backend: { callback in
                self.queue.asyncAfter(deadline: .now() + 0.12) {
                    callback(.success(BalanceSnapshot(
                        cardId: cardId,
                        balance: 38947,
                        timestamp: ISO8601DateFormatter().string(from: Date())
                    )))
                }
            },
            completion: completion
        )
    }

    func getStatus() -> SdkStatusSnapshot {
        SdkStatusSnapshot(initialized: initialized, version: "1.2.3-mock")
    }

    func shutdown() {
        initialized = false
    }

    private func handleScenarioAwareRequest<T>(
        requestKey: String,
        backend: @escaping (@escaping (Result<T, SdkErrorCode>) -> Void) -> Void,
        completion: @escaping (Result<T, SdkErrorCode>) -> Void
    ) {
        let outcome = nextOutcome(for: requestKey)

        if !outcome.emitsCallback {
            if outcome.clearAttemptsAfterHandling {
                clearAttempts(for: requestKey)
            }
            return
        }

        if !outcome.usesBackendSuccessPath {
            let errorCode = outcome.errorCode ?? .unknown
            queue.asyncAfter(deadline: .now() + 0.15) {
                completion(.failure(errorCode))
                if outcome.clearAttemptsAfterHandling {
                    self.clearAttempts(for: requestKey)
                }
            }
            return
        }

        backend { result in
            switch result {
            case .success(let value):
                completion(.success(value))
                if outcome.duplicateCallback {
                    completion(.success(value))
                }
            case .failure(let errorCode):
                completion(.failure(errorCode))
            }

            self.clearAttempts(for: requestKey)
        }
    }

    private func nextOutcome(for requestKey: String) -> ScenarioOutcome {
        let attempt = queue.sync { () -> Int in
            let nextValue = (requestAttempts[requestKey] ?? 0) + 1
            requestAttempts[requestKey] = nextValue
            return nextValue
        }
        return ScenarioOutcome.forAttempt(scenarioController.activePreset, attempt: attempt)
    }

    private func clearAttempts(for requestKey: String) {
        queue.async {
            self.requestAttempts.removeValue(forKey: requestKey)
        }
    }

    private func buildRequestKey(method: String, cardId: String, amount: Int?) -> String {
        var key = "\(method)|\(cardId.trimmingCharacters(in: .whitespacesAndNewlines))"
        if let amount {
            key += "|\(amount)"
        }
        return key
    }
}
